### PR TITLE
Support profile data parsing for tensorrtllm engine service kind

### PR DESCRIPTION
**Pull request by nv-hwoo**
- Support parsing profile export JSON from tensorrtllm engine through c api
- add tests
- small refactor for test

### CODE REVIEW

**Review by dyastremsky — COMMENTED** [review]
Great testing architecture!

Left a question. Since we may be removing service-kind, it may matter less, but it feels like the C API service-kind is being used exclusively with the direct TRT-LLM engine in mind (e.g. skipping the tokenization, getting engine token counts in a TRT-LLM-specific way). That feels like it could cause issues in the future. Should we rename the service-kind to make it clear that it's not just the C API but specifically for TRT-LLM? Customers may think this is for the general Triton C API, like what PA supports.

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 1]
```diff
@@ -410,6 +504,107 @@ def test_openai_vlm_profile_data(self, mock_read_write: pytest.MonkeyPatch) -> N
         with pytest.raises(KeyError):
             pd.get_statistics(infer_mode="concurrency", load_level="40")
 
+    def test_tensorrtllm_engine_llm_profile_data(
+        self, mock_read_write: pytest.MonkeyPatch
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:508]: gif

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 2]
```diff
@@ -410,6 +504,107 @@ def test_openai_vlm_profile_data(self, mock_read_write: pytest.MonkeyPatch) -> N
...
+            - experiment 1: [3, 3]
+            - experiment 2: [4, 3]
+        * input sequence lengths
+            - experiment 1: [3, 4]
+            - experiment 2: [3, 4]
+        """
+        tokenizer = get_tokenizer(DEFAULT_TOKENIZER)
+        pd = LLMProfileDataParser(
+            filename=Path("tensorrtllm_engine_profile_export.json"),
+            tokenizer=tokenizer,
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:539]: Let's update `LLMProfileDataParser` constructor to not require tokenizer since this particular mode (tensorrtllm_engine) does in fact not need any tokenization done by GAP.

File: genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py [hunk 3]
```diff
@@ -204,6 +204,8 @@ def _get_input_token_count(self, req_inputs: dict) -> int:
         """Deserialize the request input and return tokenized inputs."""
         if self._service_kind == "triton":
             input_text = req_inputs["text_input"]
+        elif self._service_kind == "triton_c_api":
+            return len(req_inputs["input_ids"])  # no tokenizer required
```
[the remark this answers is not in the document]
Comment (nv-hwoo) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: This is specific to TRTLLM since TRTLLM is the only model that we support through triton c api.

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 4]
```diff
@@ -410,6 +504,107 @@ def test_openai_vlm_profile_data(self, mock_read_write: pytest.MonkeyPatch) -> N
...
+        )
+
+        # experiment 1 metrics & statistics
+        stat_obj = pd.get_statistics(infer_mode="concurrency", load_level="10")
+        metrics = stat_obj.metrics
+        stat = stat_obj.stats_dict
+
+        # expected metrics
+        ttft = [2, 2]
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:548]: might help to call these `expected_tftt`

and `metrics` to `observed_metrics`

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 5]
```diff
@@ -410,6 +504,107 @@ def test_openai_vlm_profile_data(self, mock_read_write: pytest.MonkeyPatch) -> N
...
+            - experiment 2: [3, 4]
+        """
+        tokenizer = get_tokenizer(DEFAULT_TOKENIZER)
+        pd = LLMProfileDataParser(
+            filename=Path("tensorrtllm_engine_profile_export.json"),
+            tokenizer=tokenizer,
+        )
+
+        # experiment 1 metrics & statistics
+        stat_obj = pd.get_statistics(infer_mode="concurrency", load_level="10")
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:543]: We can parameterize these values and other expected values in the test in a way the reduces duplicated logic:

https://docs.pytest.org/en/7.1.x/how-to/parametrize.html

Food for thought

image

File: genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyastremsky) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: I see. I won't block on this if that's the approach that you, @matthewkotila, and @debermudez are using. It just seems misleading to make it look like this is the logic for supporting C API, but it's really just TRT-LLM via C API and our logic is exclusively designed for TRT-LLM (on the metrics side, not just the parsing side).

I would personally rather [this line](https://github.com/triton-inference-server/perf_analyzer/pull/33/files#diff-a056e8717cfab56e4f62b45b9b35efced45190c204f05ef8eea56a19bd7623e9R50) not exist and we just pass C API directly to the PA command. Then, it's clear in this file that we're only using this logic for the tensorrtllm_engine service-kind.

**Comment by nv-hwoo** [comment]
@dyastremsky atm, the service kind is coming from PA's profile export json file, and unfortunately I don't think there's an easy way to change that in PA. Also note that this is different from `args.service_kind`. From customer's perspective, they run TRTLLM engine using `--service-kind tensorrtllm_engine` so we don't actually expose to the customers that triton c api is being used. The data parser gets service kind from profile export json directly because the class is being used by `compare` subcommand where no `profile` related args are passed.

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 6]
```diff
@@ -740,3 +935,147 @@ def test_empty_response(self, mock_read_write: pytest.MonkeyPatch) -> None:
             },
         ],
     }
+
+    tensorrtllm_engine_profile_data = {
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:939]: I think it would be cleaner to define this in the function its derivative is going to be used in. There should be a way to invoke the monkey patcher after some test case code has already been run (e.g. this definition) perhaps using .context():

https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.context

File: genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyastremsky) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: Update: I think I didn't need to tag Matt or Elias here, apologies for that. I was thinking you all worked on the designs for the TRT-LLM engine path, but I'm realizing that this is simply a code question. It's not that we're adding a C API service-kind, just that we're overriding the TRT-LLM engine service-kind with C API behind the scenes (in the linked code), which led to this code being confusing on my end. I still think it could be good to not do that overriding to make the code clearer.

**Comment by dyastremsky** [comment]
> @dyastremsky Note that the service kind used in data parser is different from `args.service_kind`. From customer's perspective, they run TRTLLM engine using `--service-kind tensorrtllm_engine` so we don't actually expose to the customers that triton c api is being used. The data parser gets service kind from profile export json directly, and unfortunately I don't think there's an easy way to change that in PA. The data parser gets service kind directly from profile export json because the class is being used by `compare` subcommand where no `profile` related args are passed.

Ah, I see. Thanks! That's a bit confusing. It might be worth spending a little bit of time to see if there's a workaround or way to improve the code (maybe Matt or Elias have ideas). If not, we can keep this as is, but there's a code smell here IMO.

File: genai-perf/tests/test_llm_profile_data_parser.py [hunk 7]
```diff
@@ -41,6 +41,97 @@ def ns_to_sec(ns: int) -> Union[int, float]:
     return ns / 1e9
 
 
+def create_stats_dict(
```
Comment (matthewkotila) [inline genai-perf/tests/test_llm_profile_data_parser.py:44]: Talked offline--Hyunjae will attempt to use the pre-existing production code Statistics class to hold this information for the actual/observed data for the test.

Since no comparator functionality exists in the production code Statistics class, but one is needed for testing, a comparator function could be added just to this test file to handle that.

File: genai-perf/tests/test_llm_profile_data_parser.py (hunk 2 above)
Comment (nv-hwoo) [reply genai-perf/tests/test_llm_profile_data_parser.py:521]: talked offline -- will address this in the future PR

File: genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nv-hwoo) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: sorry could you rephrase your question? I don't think I quite understood what your suggestion was here.

[the remark this answers is not in the document]
Comment (matthewkotila) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: I think David is retracting his suggestion?

[the remark this answers is not in the document]
Comment (nv-hwoo) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: + I think what will solve a lot of confusion is another refactor in profile data parser classes so that we have one classes for each endpoints (e.g. openai chat completion, triton vllm, etc). This way, we don't have these nasty if-else blocks.

[the remark this answers is not in the document]
Comment (dyastremsky) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: Discussed offline. PA requires service-kind to be triton_c_api and GenAI-Perf uses tensorrtllm_engine. To help with clarity, Hyunjae offered to push changes to use tensorrtllm_engine as the service-kind across GenAI-Perf and just pass in triton_c_api for PA where it needs it.

[the remark this answers is not in the document]
Comment (dyastremsky) [reply genai-perf/genai_perf/profile_data_parser/llm_profile_data_parser.py:208]: Looks like the tensorrtllm_engine part is from PA's profile export and cannot be changed. Resolving.

**Review by dyastremsky — APPROVED** [review]
Spectacular work, Hyunjae! 🚀

**Review by matthewkotila — APPROVED** [review]
